modules/xfem/src/kernels/CrackTipEnrichmentStressDivergenceTensors.C

```diff
@@ -0,0 +1,237 @@
+/****************************************************************/
+
+#include "CrackTipEnrichmentStressDivergenceTensors.h"
+#include "ElasticityTensorTools.h"
+
+template <>
+InputParameters
+validParams<CrackTipEnrichmentStressDivergenceTensors>()
+{
+  InputParameters params = validParams<ALEKernel>();
+  params.addClassDescription("Enrich stress divergence kernel for the Cartesian coordinate system");
+  params.addRequiredParam<unsigned int>("component",
+                                        "An integer corresponding to the direction the variable "
+                                        "this kernel acts in. (0 for x, 1 for y, 2 for z)");
+  params.addRequiredParam<unsigned int>("enrichment_component",
+                                        "The component of the enrichement functions");
+  params.addRequiredCoupledVar("enrichment_displacements",
+                               "The string of displacements suitable for the problem statement");
+  params.addParam<std::string>("base_name", "Material property base name");
+  params.addRequiredParam<UserObjectName>("crack_front_definition",
+                                          "The CrackFrontDefinition user object name");
+  params.set<bool>("use_displaced_mesh") = false;
+  return params;
+}
+
+CrackTipEnrichmentStressDivergenceTensors::CrackTipEnrichmentStressDivergenceTensors(
+    const InputParameters & parameters)
+  : ALEKernel(parameters),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _stress(getMaterialPropertyByName<RankTwoTensor>(_base_name + "stress")),
+    _Jacobian_mult(getMaterialPropertyByName<RankFourTensor>(_base_name + "Jacobian_mult")),
+    _component(getParam<unsigned int>("component")),
+    _enrichment_component(getParam<unsigned int>("enrichment_component")),
+    _nenrich_disp(coupledComponents("enrichment_displacements")),
+    _crack_front_definition(&getUserObject<CrackFrontDefinition>("crack_front_definition")),
+    _B(4),
+    _dBX(4),
+    _dBx(4),
+    _BI(4),
+    _BJ(4)
+{
+  _enrich_disp_var.resize(_nenrich_disp);
+  for (unsigned int i = 0; i < _nenrich_disp; ++i)
+    _enrich_disp_var[i] = coupled("enrichment_displacements", i);
+}
+
+void
+CrackTipEnrichmentStressDivergenceTensors::prepareCrackTipEnrichementFunctionAtNode(
+    Node * node, std::vector<Real> & B)
+{
+  // _i-th node
+  _crack_front_definition->calculateRThetaToCrackFront(*node, _r, _theta);
+
+  Real st = std::sin(_theta);
+  Real st2 = std::sin(_theta / 2.0);
+  Real ct2 = std::cos(_theta / 2.0);
+  Real sr = std::sqrt(_r);
+
+  B[0] = sr * st2;
+  B[1] = sr * ct2;
+  B[2] = sr * st2 * st;
+  B[3] = sr * ct2 * st;
+}
+
+Real
+CrackTipEnrichmentStressDivergenceTensors::computeQpResidual()
+{
+  prepareCrackTipEnrichementFunctionAtNode(_current_elem->get_node(_i), _BI);
+
+  unsigned int crack_front_point_index =
+      _crack_front_definition->calculateRThetaToCrackFront(_q_point[_qp], _r, _theta);
+
```

Comment:
I made a new class EnrichmentFunctionCalculation which performs all the calculation the values and derivatives of those enrichment functions. And I let CrackTipEnrichmentStressDivergenceTensors and ComputeCrackTipEnrichmentSmallStrain derive from it to avoid code duplication.